### ISSUE TITLE
Pods stuck terminating/not cleaned up automatically due to invalid spec

### ISSUE THREAD

**Issue by andor44** [kubernetes/kubernetes#49492]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
Created a deployment on 1.6.4 that included pod affinity annotations. Then upgraded to 1.7.1 which changed the schema of affinity annotations. Updated the deployment with the new schema, which spun up a new RS and scaled down the old one. Pods belonging to the old RS went `Terminating` but never successfully terminated. `kubectl delete pod --force --now foo` would give a successful message but did nothing. Attempting to delete the RS would result in

`error: Scaling the resource failed with: ReplicaSet.extensions "foobar" is invalid: spec.template.annotations.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Required value: can not be empty; Current resource version 174652233
`

After editing the pods manually to match the new schema they went away immediately.

**What you expected to happen**:
The old, invalid replica set to successfully scale down and clean up its pods after the deployment is updated without needing manual intervention.

**How to reproduce it (as minimally and precisely as possible)**:
Have a deployment with a pod affinity that looks like this:
```{
  "podAntiAffinity": {
    "preferredDuringSchedulingIgnoredDuringExecution": [
      {
        "weight": 80,
        "labelSelector": {
          "matchExpressions": [
            {
              "key": "app",
              "operator": "In",
              "values": ["foobar"]
            }
          ]
        },
        "topologyKey": "failure-domain.beta.kubernetes.io/zone"
      }
    ]
  }
}
```
Needs (I assume) <1.7.0, we created it on 1.6.4.

Update to 1.7.0+, observe that pods now have events like this:
```
  FirstSeen	LastSeen	Count	From					SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----					-------------	--------	------			-------
  42m		5m		176	kubelet, xxxxxxxxxxxxxxxxxxxxxxxxxxxx			Warning		FailedValidation	Error validating pod xxx from api, ignoring: metadata.annotations.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey: Required value: can not be empty
```
Update the deployment that created the pods to match the new spec. The pods of the old replicaset will be stuck `Terminating` until the pods are manually edited to match the new schema, at which point they will clean up instantly.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): Pre-upgrade 1.6.4, post-upgrade 1.7.1
- Cloud provider or hardware configuration**: Bare metal
- OS (e.g. from /etc/os-release):
```cat /etc/os-release
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1409.7.0
VERSION_ID=1409.7.0
BUILD_ID=2017-07-19-0005
PRETTY_NAME="Container Linux by CoreOS 1409.7.0 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
```
- Kernel (e.g. `uname -a`): `uname -a
Linux xxx 4.11.11-coreos #1 SMP Tue Jul 18 23:06:59 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2630 v2 @ 2.60GHz GenuineIntel GNU/Linux`
- Install tools: Matchbox/Ansible
- Others:

**Comment by andor44** [comment]
cc @kubernetes/sig-api-machinery

**Comment by derekwaynecarr** [comment]
cc @kubernetes/sig-scheduling-misc @aveshagarwal

**Comment by aveshagarwal** [comment]
> Created a deployment on 1.6.4 that included pod affinity annotations. 
Is there any reason you are still using annotations with 1.6 and not fields? In general, I'd say it would be better if you could have migrated from annotations to fields before upgrading to 1.7. That said there was an issue (https://github.com/kubernetes/kubernetes/issues/44360) that we fixed in 1.7 which we missed in 1.6. I am still looking if your issue is related to this or not.

**Comment by wanghaoran1988** [comment]
@andor44 Could you please provide your pull deployment json/yaml file?

**Comment by aveshagarwal** [comment]
I think I can reproduce your issue by upgrading 1.6 to 1.7 and I see the same error. And it is indeed happening due to the fix to the issue https://github.com/kubernetes/kubernetes/issues/44360 .

**Comment by andor44** [comment]
>Is there any reason you are still using annotations with 1.6 and not fields?

None, other than annotations coming first so that's how these deployments were written originally.

>In general, I'd say it would be better if you could have migrated from annotations to fields before upgrading to 1.7.

The 1.7 release notes did not indicate that the annotations would stop working or break in some way so we assumed it was safe to upgrade. The note about the feature gate did not concern us as we had never enabled that feature gate. It seems to me that this is either an undocumented breaking change or a bug, hence this issue.

[Here's](https://gist.github.com/andor44/09705b275b50e8d93ef6612fe47202ff) one of the affected deployments. Be aware that this wasn't the only one.

**Comment by wanghaoran1988** [comment]
@aveshagarwal I am not clear about the problem, why the topologyKey is empty? I see the key have values:
`"topologyKey": "failure-domain.beta.kubernetes.io/zone"`

Could you give me some pointers?

**Comment by aveshagarwal** [comment]
@wanghaoran1988  my guess is that they might have updated topology key to something (so that its not empty now) during upgrade to 1.7.
The issue is that if the topology key is empty in 1.6 for pod anti affinity PreferredDuringSchedulingIgnoredDuringExecution, and then its upgraded to 1.7, it would fail because now 1.7, incorrectly, it requires the topology key to be not empty  for pod anti affinity PreferredDuringSchedulingIgnoredDuringExecution when AffinityInAnnotation is disabled.

**Comment by wanghaoran1988** [comment]
@aveshagarwal Thanks, that make sense !

**Comment by andor44** [comment]
@wanghaoran1988 @aveshagarwal no, it was not updated, it was set incorrectly. `topologyKey` needs to be under another object `podAffinityTerm`, but it was one level up. Since before 1.7 it was not mandatory it was silently ignored. We updated to 1.7 which made it mandatory, hence the failure.

**Comment by andor44** [comment]
To clarify, `preferredDuringSchedulingIgnoredDuringExecution` should have been: 
```
{
    "weight": 80,
    "labelSelector": {
        "matchExpressions": [
        {
            "key": "app",
            "operator": "In",
            "values": ["foobar"]
        }
        ]
    },
    "podAffinityTerm": {
        "topologyKey": "failure-domain.beta.kubernetes.io/zone"
    }
}
```

but it was:
```
{
    "weight": 80,
    "labelSelector": {
        "matchExpressions": [
        {
            "key": "app",
            "operator": "In",
            "values": ["foobar"]
        }
        ]
    },
    "topologyKey": "failure-domain.beta.kubernetes.io/zone"
}
```

Omitting `topologyKey` is no longer permitted in 1.7, therefore the second one became invalid.

**Comment by aveshagarwal** [comment]
@andor44 After further discussion with sig scheduling group, it is intended to not allow empty topology key with soft pod anti affinity (preferredDuringSchedulingIgnoredDuringExecution pod anti affinity). So it is working as expected. So i think this bug could be closed.

**Comment by andor44** [comment]
@aveshagarwal I realize at this point that the name of the issue is probably misleading. My issue is not that the topology key is required. I understand that it is required, I am not reporting that as the bug. The bug is that I had an object already Kubernetes that became invalid after the upgrade. Once it became invalid it stayed in Kubernetes and it could not be deleted. Delete requests returned a success return code, but the object didn't go away. It could only be deleted once the object was manually edited to conform to the spec. This behavior is very counter-intuitive and I do not see the reason for disallowing object deletion of invalid objects.

Should I edit the title, or open a new bug and reference this one?

**Comment by aveshagarwal** [comment]
@andor44 ok then you could keep it open.

**Comment by alex88** [comment]
In my case, having an old replicaset with a bad configuration doesn't let me create new replicasets so deploy new versions as old replicaset still has a desired number of pods that can't run due configuration errors

**Comment by alex88** [comment]
Even if deployment has no affinity configuration, doing a `get deployment -o yaml` shows that annotation `kubectl.kubernetes.io/last-applied-configuration` has that affinity config and status shows this failure
```
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: 2017-08-25T00:37:32Z
    lastUpdateTime: 2017-08-25T00:37:32Z
    message: 'unable to create pods: Pod "php-2221189365-kzxvk" is invalid: metadata.annotations.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0].podAffinityTerm.topologyKey:
      Required value: can not be empty'
    reason: FailedCreate
    status: "True"
    type: ReplicaFailure
```

**Comment by ravisantoshgudimetla** [comment]
@andor44 - Can you please change the labels as this issue is not related to sig-scheduling anymore? I think adding appropriate labels might help in getting the attention of sig's

**Comment by andor44** [comment]
I don't know what sig is appropriate, maybe API machinery?

**Comment by 3h4x** [comment]
Just run into this issue after upgrading k8s from 1.6.6 to 1.7.12.
Deployments working in 1.6.6 stopped working after upgrade and couldn't be delete even when `topologyKey` was added.

**Comment by martynd** [comment]
Hit this also, 1.6.13 to 1.7.12

Worked around the issues though.

For those still affected:
TLDR; Fix the deployment if you haven't already, then edit the latest (pre upgrade) replicaset (it should be the 2nd newest, with the newest being the one just created). Once the replicaset is correct, the rollout is happy and cleans things up


The reason editing the deployment doesn't fix it (and leaves old pods in a strange state) is the old replicaset.

When the deployment is edited in an attempt to add the topologykey, it makes a new replicaset and scales down the old. 
As the old one is missing the key, the internal rollout process fails to scale down the number of replicas as the replicaset is invalid. 
From here it all just gets stuck with little explanation as the deployment is now referring to the new correct replicaset
